File: code/python/data_structures_article-13.py
"""Design the timeline function of Moments.

Think of every tweet as a card in a linked chain (newest first).
The feed pulls the newest visible card and then updates to the next newest one.
"""

from __future__ import annotations

from heapq import heappush, heappop
from typing import Any, Dict, List


def solve(*args: Any, **kwargs: Any) -> Any:
    """Build and return a Twitter timeline object for quick learning."""
    return Twitter(*args, **kwargs)


class Twitter:
    def __init__(self, capacity: int):
        self.globalTime = 0
        self.idToUser: Dict[int, "Twitter.User"] = {}

    class Tweet:
        def __init__(self, id: int, owner: "Twitter"):
            self.owner = owner
            self.id = id
            self.timestamp = owner.globalTime
            owner.globalTime += 1
            self.next = None

        def get_id(self) -> int:
            return self.id

        def get_timestamp(self) -> int:
            return self.timestamp

        def get_next(self):
            return self.next

        def set_next(self, next_tweet):
            self.next = next_tweet

    class User:
        def __init__(self, id: int):
            self.id = id
            self.tweet_head = None
            self.followed_user_set = set()

        def get_id(self):
            return self.id

        def get_tweet_head(self):
            return self.tweet_head

        def get_followed_user_set(self):
            return self.followed_user_set

        def __eq__(self, other):
            return self.id == other.id

        def follow(self, other):
            self.followed_user_set.add(other)

        def unfollow(self, other):
            self.followed_user_set.discard(other)

        def post(self, tweet: "Twitter.Tweet"):
            tweet.set_next(self.tweet_head)
            self.tweet_head = tweet

    def postTweet(self, userId: int, tweetId: int) -> None:
        if userId not in self.idToUser:
            self.idToUser[userId] = self.User(userId)
        user = self.idToUser[userId]
        tweet = Twitter.Tweet(tweetId, self)
        user.post(tweet)

    def getNewsFeed(self, userId: int) -> List[int]:
        res: List[int] = []
        if userId not in self.idToUser:
            return res

        user = self.idToUser[userId]
        followed_user_set = user.get_followed_user_set()
        pq = []

        if user.get_tweet_head():
            heappush(pq, (-user.tweet_head.timestamp, user.get_tweet_head()))

        for other in followed_user_set:
            if other.get_tweet_head():
                heappush(pq, (-other.tweet_head.timestamp, other.get_tweet_head()))

        count = 0
        while pq and count < 10:
            _, tweet = heappop(pq)
            res.append(tweet.get_id())
            nxt = tweet.get_next()
            if nxt:
                heappush(pq, (-nxt.timestamp, nxt))
            count += 1
        return res

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId not in self.idToUser:
            self.idToUser[followerId] = self.User(followerId)
        if followeeId not in self.idToUser:
            self.idToUser[followeeId] = self.User(followeeId)
        follower = self.idToUser[followerId]
        followee = self.idToUser[followeeId]
        follower.follow(followee)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followerId not in self.idToUser or followeeId not in self.idToUser:
            return
        follower = self.idToUser[followerId]
        followee = self.idToUser[followeeId]
        follower.unfollow(followee)

File: code/java/data_structures_article-13.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

/**
 * Design the timeline of moments with a tiny social feed:
 * each user has tweets in a linked list, and we always pull the newest tweet next.
 */
public class Data_structures_article_13 {

    /**
     * Build the Twitter simulator by giving the number of recent tweets to show.
     */
    public Object solve(Object... args) {
        if (args == null || args.length != 1) {
            throw new IllegalArgumentException("Expected: (int userCount)");
        }

        int cap = toInt(args[0]);
        return new Twitter(cap);
    }

    private int toInt(Object value) {
        if (value instanceof Number) {
            return ((Number) value).intValue();
        }
        return Integer.parseInt(String.valueOf(value));
    }

    private static class Twitter {
        int globalTime = 0;
        Map<Integer, User> idToUser = new HashMap<>();

        class Tweet {
            private final int id;
            private final int timestamp;
            private Tweet next;

            public Tweet(int id) {
                this.id = id;
                this.timestamp = globalTime++;
            }
            public int getId() {
                return id;
            }
            public int getTimestamp() {
                return timestamp;
            }
            public Tweet getNext() {
                return next;
            }
            public void setNext(Tweet next) {
                this.next = next;
            }
        }

        class User {
            private final int id;
            private Tweet tweetHead;
            private final Set<User> followedUserSet;

            public User(int id) {
                this.id = id;
                this.tweetHead = null;
                this.followedUserSet = new HashSet<>();
            }
            public int getId() {
                return id;
            }
            public Tweet getTweetHead() {
                return tweetHead;
            }
            public Set<User> getFollowedUserSet() {
                return followedUserSet;
            }
            public boolean equals(User other) {
                return this.id == other.id;
            }
            public void follow(User other) {
                followedUserSet.add(other);
            }
            public void unfollow(User other) {
                followedUserSet.remove(other);
            }
            public void post(Tweet tweet) {
                tweet.setNext(tweetHead);
                tweetHead = tweet;
            }
        }

        public Twitter(int capacity) {
            this.globalTime = 0;
        }

        public void postTweet(int userId, int tweetId) {
            if (!idToUser.containsKey(userId)) {
                idToUser.put(userId, new User(userId));
            }
            User user = idToUser.get(userId);
            user.post(new Tweet(tweetId));
        }

        public int[] getNewsFeed(int userId) {
            if (!idToUser.containsKey(userId)) {
                return new int[]{};
            }

            User user = idToUser.get(userId);
            Set<User> followedUserSet = user.getFollowedUserSet();

            PriorityQueue<Tweet> pq = new PriorityQueue<>((a, b) -> {
                return b.timestamp - a.timestamp;
            });

            if (user.getTweetHead() != null) {
                pq.offer(user.getTweetHead());
            }
            for (User other : followedUserSet) {
                if (other.getTweetHead() != null) {
                    pq.offer(other.tweetHead);
                }
            }

            int[] res = new int[10];
            int count = 0;
            while (!pq.isEmpty() && count < 10) {
                Tweet tweet = pq.poll();
                res[count++] = tweet.getId();
                if (tweet.getNext() != null) {
                    pq.offer(tweet.getNext());
                }
            }
            if (count == 10) {
                return res;
            }
            int[] shortRes = new int[count];
            System.arraycopy(res, 0, shortRes, 0, count);
            return shortRes;
        }

        public void follow(int followerId, int followeeId) {
            if (!idToUser.containsKey(followerId)) {
                idToUser.put(followerId, new User(followerId));
            }
            if (!idToUser.containsKey(followeeId)) {
                idToUser.put(followeeId, new User(followeeId));
            }
            User follower = idToUser.get(followerId);
            User followee = idToUser.get(followeeId);
            follower.follow(followee);
        }

        public void unfollow(int followerId, int followeeId) {
            if (!idToUser.containsKey(followerId) || !idToUser.containsKey(followeeId)) {
                return;
            }
            User follower = idToUser.get(followerId);
            User followee = idToUser.get(followeeId);
            follower.unfollow(followee);
        }
    }
}
